llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp

```diff
@@ -2327,11 +2327,15 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
 #endif
 
     if (ST->isPreciseMemoryEnabled() && Inst.mayLoadOrStore()) {
-      AMDGPU::Waitcnt Wait = WCG->getAllZeroWaitcnt(
-          Inst.mayStore() && !SIInstrInfo::isAtomicRet(Inst));
-      ScoreBrackets.simplifyWaitcnt(Wait);
-      Modified |= generateWaitcnt(Wait, std::next(Inst.getIterator()), Block,
-                                  ScoreBrackets, /*OldWaitcntInstr=*/nullptr);
+      Iter++;
+      if (!isWaitInstr(*Iter)) {
+        AMDGPU::Waitcnt Wait = WCG->getAllZeroWaitcnt(
+            Inst.mayStore() && !SIInstrInfo::isAtomicRet(Inst));
+        ScoreBrackets.simplifyWaitcnt(Wait);
+        Modified |= generateWaitcnt(Wait, std::next(Inst.getIterator()), Block,
+                                    ScoreBrackets, /*OldWaitcntInstr=*/nullptr);
+      }
+      Iter--;
```

Comment:
You shouldn't need to manipulate the iterator here. Do you just need to get rid of the std::next, or use Iter directly above?

Reply:
When I see a load/store instruction, I need to check if the next one is a waitcnt. A waitcnt is only inserted if the next one is not a waitcnt. Thus the Iter++ and Iter--.

Reply:
Something is wrong, you shouldn't need to check for neighbors?